From a85ab706457748a4afc982f63cea6f8c41977b78 Mon Sep 17 00:00:00 2001
From: Mark Janes <markjanes@swizzler.org>
Date: Tue, 24 May 2022 12:16:27 -0700
Subject: [PATCH] intel: parse intel_measure environment without side effects

If an application links agaist both iris and anv, they will clash when
parsing the INTEL_MEASURE environment variable.

Reviewed-by: Kenneth Graunke <kenneth@whitecape.org>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/16571>
(cherry picked from commit 3c53c6b247b32927bb18f5be526bdf11db13d745)

Also resolves a C23/glibc build error: with C23 glibc, strrchr() returns
const char * when given a const char * argument, so "char *sep = strrchr(env,
',')" fails to compile under -Werror=incompatible-pointer-types-discards-
qualifiers. Copying env into a local mutable char env_copy[1024] makes
strrchr() return char * and also removes the (previously UB) write to the
getenv() string.

[Backported to 22.0.2]
Signed-off-by: Adrian Ratiu <adrian.ratiu@collabora.com>
---
 src/intel/common/intel_measure.c | 22 +++++++++++++---------
 1 file changed, 13 insertions(+), 9 deletions(-)

diff --git a/src/intel/common/intel_measure.c b/src/intel/common/intel_measure.c
index c8474ac3809..12f0d719e24 100644
--- a/src/intel/common/intel_measure.c
+++ b/src/intel/common/intel_measure.c
@@ -64,8 +64,12 @@ intel_measure_init(struct intel_measure_device *device)
       if (!env)
          return;
 
+      char env_copy[1024];
+      strncpy(env_copy, env, 1024);
+      env_copy[1023] = '\0';
+
       config.file = stderr;
-      config.flags = parse_debug_string(env, debug_control);
+      config.flags = parse_debug_string(env_copy, debug_control);
       if (!config.flags)
          config.flags = INTEL_MEASURE_DRAW;
       config.enabled = true;
@@ -87,15 +91,15 @@ intel_measure_init(struct intel_measure_device *device)
       const int DEFAULT_BUFFER_SIZE = 16 * 1024;
       config.buffer_size = DEFAULT_BUFFER_SIZE;
 
-      const char *filename = strstr(env, "file=");
-      const char *start_frame_s = strstr(env, "start=");
-      const char *count_frame_s = strstr(env, "count=");
-      const char *control_path = strstr(env, "control=");
-      const char *interval_s = strstr(env, "interval=");
-      const char *batch_size_s = strstr(env, "batch_size=");
-      const char *buffer_size_s = strstr(env, "buffer_size=");
+      const char *filename = strstr(env_copy, "file=");
+      const char *start_frame_s = strstr(env_copy, "start=");
+      const char *count_frame_s = strstr(env_copy, "count=");
+      const char *control_path = strstr(env_copy, "control=");
+      const char *interval_s = strstr(env_copy, "interval=");
+      const char *batch_size_s = strstr(env_copy, "batch_size=");
+      const char *buffer_size_s = strstr(env_copy, "buffer_size=");
       while (true) {
-         char *sep = strrchr(env, ',');
+         char *sep = strrchr(env_copy, ',');
          if (sep == NULL)
             break;
          *sep = '\0';
-- 
2.53.0

